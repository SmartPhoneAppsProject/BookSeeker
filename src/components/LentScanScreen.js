import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  Image,
  Dimensions,
} from 'react-native';
import {
  BarCodeScanner,
  Permissions
} from 'expo';
import { Button } from 'react-native-elements';

import { rentBook } from '../utils/Network';
import { NavigationActions } from "react-navigation";

export default class lentScanScreen extends Component {
  static navigationOptions = {
    title: '貸し出し',
  };

  constructor(props) {
    super(props);
    this.state = {
      janCode: null,
      permissionsGranted: false,
      status: 'reading',
    };
  }

  async componentWillMount() {
    const { status } = await Permissions.askAsync(Permissions.CAMERA);
    this.setState({ permissionsGranted: status === 'granted' });
  }

  goToHomeScreen = () => {
    // https://github.com/react-navigation/react-navigation/issues/1448
    const actions = [NavigationActions.navigate({ routeName: 'Home' })];

    const resetAction = NavigationActions.reset({
      index: actions.length - 1,
      actions
    });

    this.props.navigation.dispatch(resetAction)
  };

  lendBook = (janCode) => {
    const json = JSON.stringify({
      jan_code: janCode,
      status: true
    });

    this.changeBookStatus(json);
  };

  returnBook = (janCode) => {
    const json = JSON.stringify({
      jan_code: janCode,
      status: false
    });

    this.changeBookStatus(json);
  };

  changeBookStatus = (json) => {
    rentBook(json)
      .then(response => response.json())
      .then(responseJson => {
        console.log(responseJson);
      })
      .catch(error => {
        console.warn(error);
        rentBook(json)
          .then(response => response.json())
          .then(responseJson => {
            console.log(responseJson);
          })
          .catch(error => console.error(error));
      });
  };

  _handleBarCodeRead = ({ type, data }) => {
    const { action } = this.props.navigation.state.params;

    console.log(data);
    if (`${BarCodeScanner.Constants.BarCodeType.ean13}` == type) {
      if (978 == data.slice(0, 3)) { //ISBNを読み取ったとき
        if (this.state.janCode != data) {
          this.setState({
            janCode: data,
            status: 'ok'
          });
          const janCode = parseInt(data);
          if (action == 'return') {
            this.returnBook(janCode);
          } else if (action == 'lend') {
            this.lendBook(janCode);
          }
          this.goToHomeScreen();
        }
      } else { //バーコードであるがISBNでないとき
        this.setState({ status: 'invalid' });
      }
      setTimeout(() => {
        this.setState({ status: 'reading' });
      }, 1000);
    } else { //バーコードでないとき
      this.setState({ status: 'reading' });
    }
  };

  renderNoPermissions = () => {
    return (
      <View style={styles.noPermissions}>
        <Text style={styles.noPermissionsText}>
          カメラを使用できません
        </Text>
      </View>
    );
  };

  renderHeader = () => {
    return (
      <View style={styles.header}>
        <Text style={styles.stringWarn}>
          978
          <Text style={styles.text}>
            から始まるバーコードを画面に合わせてください
          </Text>
        </Text>
      </View>
    );
  };

  renderCamera = () => {
    return (
      <View style={styles.cameraContainer}>
        <BarCodeScanner
          style={styles.camera}
          onBarCodeRead={this._handleBarCodeRead}
        >
          <View style={styles.cameraInline}>
            <Text/>
          </View>
        </BarCodeScanner>
      </View>
    );
  };

  renderFooter = () => {
    let statusText = <View/>;
    if (this.state.status === 'ok') {
      statusText = <Text style={styles.statusOk}>読み取りました</Text>;
    } else if (this.state.status === 'invalid') {
      statusText = <Text style={styles.statusNo}>数字をお確かめください</Text>;
    }

    return (
      <View style={styles.footer}>
        <Button
          title={statusText}
          loading={(this.state.status === 'reading')}
          loadingProps={{ size: "large", color: "rgba(111, 202, 186, 1)" }}
          titleStyle={{ fontWeight: "700" }}
          clear={true}
        />
      </View>
    );
  };

  renderScanScreen = () => {
    const header = this.renderHeader();
    const camera = this.renderCamera();
    const footer = this.renderFooter();

    return (
      <View style={styles.cameraScreen}>
        {header}
        <Image
          style={styles.imageSize}
          source={require('../../assets/ISBN_sample.png')}
        />
        {camera}
        {footer}
      </View>
    );
  };

  render() {
    const screen = this.state.permissionsGranted
      ? this.renderScanScreen()
      : this.renderNoPermissions();

    return (
      <View style={styles.container}>
        {screen}
      </View>
    );
  }
}

const { width } = Dimensions.get('window');

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#000',
  },
  noPermissions: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
  },
  noPermissionsText: {
    color: 'white'
  },
  cameraScreen: {
    flex: 1,
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  header: {
    alignItems: 'center',
    margin: 5,
  },
  stringWarn: {
    color: '#cd5c5c',
  },
  text: {
    color: '#f3f3f3',
    textAlign: 'center',
  },
  imageSize: {
    width: 221 / 2,
    height: 93 / 2
  },
  body: {
    alignItems: 'center',
    margin: 5,
  },
  cameraContainer: {
    width: width * 2 / 3,
    height: width * 2 / 3,
    borderColor: '#f3f3f3',
    borderWidth: 1,
  },
  camera: {
    flex: 1,
  },
  cameraInline: {
    flex: 0.5,
    borderColor: '#cd5c5c',
    borderBottomWidth: 1,
  },
  footer: {
    margin: 5,
  },
  statusOk: {
    textAlign: 'center',
    fontSize: 30,
    color: '#3eb370'
  },
  statusNo: {
    textAlign: 'center',
    fontSize: 30,
    color: '#e95464'
  },
});
