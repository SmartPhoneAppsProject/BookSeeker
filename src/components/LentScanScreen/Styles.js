import {
  Dimensions,
  StyleSheet,
} from 'react-native';

const { width } = Dimensions.get('window');

export const index = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#000',
  },
  noPermissions: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
  },
  noPermissionsText: {
    color: 'white',
  },
  cameraScreen: {
    flex: 1,
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  header: {
    alignItems: 'center',
    margin: 5,
  },
  stringWarn: {
    color: '#cd5c5c',
  },
  text: {
    color: '#f3f3f3',
    textAlign: 'center',
  },
  imageSize: {
    width: 221 / 2,
    height: 93 / 2,
  },
  body: {
    alignItems: 'center',
    margin: 5,
  },
  cameraContainer: {
    width: width * (2 / 3),
    height: width * (2 / 3),
    borderColor: '#f3f3f3',
    borderWidth: 1,
  },
  camera: {
    flex: 1,
  },
  cameraInline: {
    flex: 0.5,
    borderColor: '#cd5c5c',
    borderBottomWidth: 1,
  },
  footer: {
    margin: 5,
  },
  statusOk: {
    textAlign: 'center',
    fontSize: 30,
    color: '#3eb370',
  },
  statusNo: {
    textAlign: 'center',
    fontSize: 30,
    color: '#e95464',
  },
});
