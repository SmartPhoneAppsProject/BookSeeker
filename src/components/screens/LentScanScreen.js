import React, { Component } from 'react';
import {
  StyleSheet,
  View,
  Image,
  Dimensions,
  ActivityIndicator,
} from 'react-native';
import {
  BarCodeScanner,
  Permissions,
} from 'expo';
import { Text } from 'react-native-elements';

import { rentBook } from '../../utils/Network';
import { navigate } from '../../utils/NavigationService';

export default class lentScanScreen extends Component {
  static navigationOptions = {
    title: '貸し出し',
  };

  constructor(props) {
    super(props);
    this.state = {
      janCode: null,
      permissionsGranted: false,
      status: 'reading',
    };
  }

  async componentWillMount() {
    const { status } = await Permissions.askAsync(Permissions.CAMERA);
    this.setState({ permissionsGranted: status === 'granted' });
  }

  lendBook = (janCode) => {
    const json = JSON.stringify({
      jan_code: janCode,
      status: true,
    });

    this.changeBookStatus(json);
  };

  returnBook = (janCode) => {
    const json = JSON.stringify({
      jan_code: janCode,
      status: false,
    });

    this.changeBookStatus(json);
  };

  changeBookStatus = (json) => {
    rentBook(json)
      .then(response => response.json())
      .then((responseJson) => {
        console.log(responseJson);
      })
      .catch((error) => {
        console.warn(error);
        rentBook(json)
          .then(response => response.json())
          .then((responseJson) => {
            console.log(responseJson);
          })
          .catch(e => console.error(e));
      });
  };

  handleBarCodeRead = ({ type, data }) => {
    const { action } = this.props.navigation.state.params;

    console.log(data);
    if (BarCodeScanner.Constants.BarCodeType.ean13 === type) {
      if (data.slice(0, 3) === '978') { // ISBNを読み取ったとき
        if (this.state.janCode !== data) {
          this.setState({
            janCode: data,
            status: 'ok',
          });
          const janCode = parseInt(data, 10);
          if (action === 'return') {
            this.returnBook(janCode);
          } else if (action === 'lend') {
            this.lendBook(janCode);
          }
          navigate('Home');
        }
      } else { // バーコードであるがISBNでないとき
        this.setState({ status: 'invalid' });
      }
      setTimeout(() => {
        this.setState({ status: 'reading' });
      }, 1000);
    } else { // バーコードでないとき
      this.setState({ status: 'reading' });
    }
  };

  renderNoPermissions = () => (
    <View style={styles.noPermissions}>
      <Text style={styles.noPermissionsText}>
        カメラを使用できません
      </Text>
    </View>
  );

  renderHeader = () => (
    <View style={styles.header}>
      <Text style={styles.stringWarn}>
        978
        <Text style={styles.text}>
          から始まるバーコードを画面に合わせてください
        </Text>
      </Text>
    </View>
  );

  renderCamera = () => (
    <View style={styles.cameraContainer}>
      <BarCodeScanner
        style={styles.camera}
        onBarCodeRead={this.handleBarCodeRead}
      >
        <View style={styles.cameraInline}>
          <Text />
        </View>
      </BarCodeScanner>
    </View>
  );

  renderFooter = () => {
    let statusText = <ActivityIndicator size="large" />;
    if (this.state.status === 'ok') {
      statusText = <Text h4 style={styles.statusOk}>読み取りました</Text>;
    } else if (this.state.status === 'invalid') {
      statusText = <Text h4 style={styles.statusNo}>数字をお確かめください</Text>;
    }

    return (
      <View style={styles.footer}>
        {statusText}
      </View>
    );
  };

  renderScanScreen = () => {
    const header = this.renderHeader();
    const camera = this.renderCamera();
    const footer = this.renderFooter();

    return (
      <View style={styles.cameraScreen}>
        {header}
        <Image
          style={styles.imageSize}
          source={require('../../../assets/ISBN_sample.png')}
        />
        {camera}
        {footer}
      </View>
    );
  };

  render() {
    const screen = this.state.permissionsGranted
      ? this.renderScanScreen()
      : this.renderNoPermissions();

    return (
      <View style={styles.container}>
        {screen}
      </View>
    );
  }
}

const { width } = Dimensions.get('window');

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#000',
  },
  noPermissions: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
  },
  noPermissionsText: {
    color: 'white',
  },
  cameraScreen: {
    flex: 1,
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  header: {
    alignItems: 'center',
    margin: 5,
  },
  stringWarn: {
    color: '#cd5c5c',
  },
  text: {
    color: '#f3f3f3',
    textAlign: 'center',
  },
  imageSize: {
    width: 221 / 2,
    height: 93 / 2,
  },
  body: {
    alignItems: 'center',
    margin: 5,
  },
  cameraContainer: {
    width: width * (2 / 3),
    height: width * (2 / 3),
    borderColor: '#f3f3f3',
    borderWidth: 1,
  },
  camera: {
    flex: 1,
  },
  cameraInline: {
    flex: 0.5,
    borderColor: '#cd5c5c',
    borderBottomWidth: 1,
  },
  footer: {
    margin: 5,
  },
  statusOk: {
    textAlign: 'center',
    fontSize: 30,
    color: '#3eb370',
  },
  statusNo: {
    textAlign: 'center',
    fontSize: 30,
    color: '#e95464',
  },
});
