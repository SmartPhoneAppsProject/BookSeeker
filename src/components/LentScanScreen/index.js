import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  View,
  Image,
  ActivityIndicator,
} from 'react-native';
import {
  BarCodeScanner,
  Permissions,
} from 'expo';
import { Text } from 'react-native-elements';

import { index as styles } from './Styles';
import { rentBook } from '../../utils/Network';
import { navigate } from '../../utils/NavigationService';

export default class LentScanScreen extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isbn: null,
      permissionsGranted: false,
      status: 'reading',
    };
  }

  async componentWillMount() {
    const { status } = await Permissions.askAsync(Permissions.CAMERA);
    this.setState({ permissionsGranted: status === 'granted' });
  }

  borrowBook = (isbn) => {
    const json = JSON.stringify({
      jan_code: isbn,
      status: true,
    });

    this.changeBookStatus(json);
  };

  returnBook = (isbn) => {
    const json = JSON.stringify({
      jan_code: isbn,
      status: false,
    });

    this.changeBookStatus(json);
  };

  changeBookStatus = (json) => {
    rentBook(json)
      .then(response => response.json())
      .then((responseJson) => {
      })
      .catch((error) => {
        console.warn(error);
        rentBook(json)
          .then(response => response.json())
          .then((responseJson) => {
            console.log(responseJson);
          })
          .catch(e => console.error(e));
      });
  };

  handleBarCodeRead = ({ type, data }) => {
    const { action } = this.props;

    if (BarCodeScanner.Constants.BarCodeType.ean13 === type) {
      if (data.slice(0, 3) === '978') { // ISBNを読み取ったとき
        if (this.state.isbn !== data) {
          this.setState({
            isbn: data,
            status: 'ok',
          });
          const isbn = parseInt(data, 10);
          if (action === 'return') {
            this.returnBook(isbn);
          } else if (action === 'borrow') {
            this.borrowBook(isbn);
          }
          navigate('Home');
        }
      } else { // バーコードであるがISBNでないとき
        this.setState({ status: 'invalid' });
      }
      setTimeout(() => {
        this.setState({ status: 'reading' });
      }, 1000);
    } else { // バーコードでないとき
      this.setState({ status: 'reading' });
    }
  };

  renderNoPermissions = () => (
    <View style={styles.noPermissions}>
      <Text style={styles.noPermissionsText}>
        カメラを使用できません
      </Text>
    </View>
  );

  renderHeader = () => (
    <View style={styles.header}>
      <Text style={styles.stringWarn}>
        978
        <Text style={styles.text}>
          から始まるバーコードを画面に合わせてください
        </Text>
      </Text>
    </View>
  );

  renderCamera = () => (
    <View style={styles.cameraContainer}>
      <BarCodeScanner
        style={styles.camera}
        onBarCodeRead={this.handleBarCodeRead}
      >
        <View style={styles.cameraInline}>
          <Text />
        </View>
      </BarCodeScanner>
    </View>
  );

  renderFooter = () => {
    let statusText = <ActivityIndicator size="large" />;
    if (this.state.status === 'ok') {
      statusText = <Text h4 style={styles.statusOk}>読み取りました</Text>;
    } else if (this.state.status === 'invalid') {
      statusText = <Text h4 style={styles.statusNo}>数字をお確かめください</Text>;
    }

    return (
      <View style={styles.footer}>
        {statusText}
      </View>
    );
  };

  renderScanScreen = () => {
    const header = this.renderHeader();
    const camera = this.renderCamera();
    const footer = this.renderFooter();

    return (
      <View style={styles.cameraScreen}>
        {header}
        <Image
          style={styles.imageSize}
          source={require('../../../assets/ISBN_sample.png')}
        />
        {camera}
        {footer}
      </View>
    );
  };

  render() {
    const screen = this.state.permissionsGranted
      ? this.renderScanScreen()
      : this.renderNoPermissions();

    return (
      <View style={styles.container}>
        {screen}
      </View>
    );
  }
}

LentScanScreen.navigationOptions = {
  title: '貸し出し',
};

LentScanScreen.propTypes = {
  action: PropTypes.oneOf(['borrow', 'return']).isRequired,
};
