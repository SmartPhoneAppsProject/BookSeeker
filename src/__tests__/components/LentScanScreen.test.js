import React from 'react';
import renderer from 'react-test-renderer';
import _book from '../../api/books.json';

import LentScanScreen from '../../components/LentScanScreen';

describe('<LentScanScreen />', () => {
  it('正しくレンダリングされること', () => {
    const book = _book[0];
    const tree = renderer
      .create(<LentScanScreen
        bookStatus
        permissions="granted"
        cameraStatus="reading"
        jancode={book.jancode}
        permissionsGranted={jest.fn()}
        permissionsDenied={jest.fn()}
        jancodeReading={jest.fn()}
        jancodeOk={jest.fn()}
        jancodeInvalid={jest.fn()}
        requestChangeStatus={jest.fn()}
        changeStatusFromJancode={jest.fn()}
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
