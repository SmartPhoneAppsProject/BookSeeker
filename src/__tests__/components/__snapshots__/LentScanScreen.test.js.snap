// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LentScanScreen /> 正しくレンダリングされること 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "space-around",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "margin": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {},
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            Object {
              "color": "#cd5c5c",
            },
          ]
        }
      >
        978
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Object {
                "color": "#f3f3f3",
                "textAlign": "center",
              },
            ]
          }
        >
          から始まるバーコードを画面に合わせてください
        </Text>
      </Text>
    </View>
    <Image
      source="test-file-stub"
      style={
        Object {
          "height": 46.5,
          "width": 110.5,
        }
      }
    />
    <View
      style={
        Object {
          "borderColor": "#f3f3f3",
          "borderWidth": 1,
          "height": 500,
          "width": 500,
        }
      }
    >
      <ExponentCamera
        autoFocus={undefined}
        barCodeScannerEnabled={true}
        barCodeTypes={Array []}
        faceDetectionClassifications={undefined}
        faceDetectionLandmarks={undefined}
        faceDetectionMode={undefined}
        flashMode={undefined}
        focusDepth={0}
        onBarCodeRead={[Function]}
        onCameraReady={[Function]}
        onFacesDetected={[Function]}
        onMountError={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
        type={undefined}
        whiteBalance={undefined}
        zoom={0}
      >
        <View
          style={
            Object {
              "borderBottomWidth": 1,
              "borderColor": "#cd5c5c",
              "flex": 0.5,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {},
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
              ]
            }
          />
        </View>
      </ExponentCamera>
    </View>
    <View
      style={
        Object {
          "margin": 5,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#999999"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
  </View>
</View>
`;
